File: leetcode-76-Minimum_Window_Substring/leetcode-76-Minimum_Window_Substring.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <unordered_map>


class Solution {
public:
    std::string minWindow(std::string s, std::string t) {
        size_t letters_c = t.size(), str_c = s.size();
        if(str_c<letters_c || s.empty() || t.empty()){
            return "";
        }

        std::unordered_map<char, size_t> letters;
        for(const auto& x:t)
        {
            ++letters[x];
        }

        // -------------------------------
        
        size_t left = 0, right = 0,
               min_len = INT_MAX, min_left = 0,
               count_l = letters_c;
        std::unordered_map<char, size_t> chack_l;

        while(right<str_c)
        {
            while(count_l)
            {
                if(right<str_c)
                {
                    if(letters.find(s[right]) != letters.end())
                    {
                        if((chack_l[s[right]]++) < letters[s[right]])
                        {
                            --count_l;
                        }
                        if(chack_l[s[right]] == letters[s[right]] &&!(count_l))
                        {
                            ++right;
                            break;
                        }
                    }
                } else {
                    break;
                }
                ++right;
            }
            if(((right-left)<min_len) && !count_l)
            {
                min_len = (right-left);
                min_left = left;
            }
            while(!count_l)
            {
                char a = s[left];
                if(letters.find(a) != letters.end())
                {
                    if(--chack_l[a] < letters[a])
                    {
                        ++count_l;
                        ++left;
                        break;
                    }
                }
                if(str_c<(++left))
                {
                    break;
                }
                if((right-left)<min_len)
                {
                    min_len = (right-left);
                    min_left = left;
                }
            }
        }
        return (min_len == INT_MAX)?(""):(s.substr(min_left,min_len));
    }
};


int main(){
    std::string s = "ADOBECODEBANC",
                t = "ABC";

    auto res = Solution().minWindow(s, t);
    
    std::cout << res << std::endl;
    
    return EXIT_SUCCESS;
}

File: leetcode-76-Minimum_Window_Substring/leetcode-76-Minimum_Window_Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        t_characters_counter: dict[str, int] = {}
        for c in t:
            if c not in t_characters_counter:
                t_characters_counter[c] = 0
            t_characters_counter[c] += 1
        
        min_window: str = ""
        min_len: float = float("inf")
        window_len: int = 0
        left: int = 0
        for right, value in enumerate(s):
            if value in t_characters_counter:
                t_characters_counter[value] -= 1
                if t_characters_counter[value] >= 0:
                    window_len += 1
            while left <= right and window_len == len(t):
                if min_len > right-left+1:
                    min_len = right-left+1
                    min_window = s[left:right+1:]
                if s[left] in t_characters_counter:
                    t_characters_counter[s[left]] += 1
                    if t_characters_counter[s[left]] > 0:
                        window_len -= 1
                left += 1
        return min_window


if __name__ == "__main__":
    s: str = "ADOBECODEBANC"
    t: str = "ABC"
    result = Solution().minWindow(s, t)
    print(result)
